File: leetcode/2024_7_july/python/16_step_by_step_directions_from_one_binary_tree_node_to_another.py
# CODE

from typing import List, Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def getDirections(self, root: Optional[TreeNode], startValue: int, destValue: int) -> str:
        
        # function to find root to node path
        def root_to_node_path(root: TreeNode, target: int, path: List[str]):
            if root is None:
                return False

            if root.val == target:
                return True
            
            # go left
            path.append('L')
            if root_to_node_path(root.left, target, path):
                return True
            path.pop()

            # go right
            path.append('R')            
            if root_to_node_path(root.right, target, path):
                return True
            path.pop()

            return False

        start_node_path = []
        end_node_path = []

        root_to_node_path(root, startValue, start_node_path)
        root_to_node_path(root, destValue, end_node_path)

        directions = []
        i = 0
        while(i < len(start_node_path) and 
                i < len(end_node_path) and 
                start_node_path[i] == end_node_path[i]):

                i += 1
        
        directions.extend('U' * (len(start_node_path) - i))
        directions.extend(end_node_path[i:])
        return "".join(directions)

File: leetcode/2024_7_july/c++/16_step_by_step_directions_from_one_binary_tree_node_to_another.cpp
// TIME COMPLEXITY
// O(n)

// SPACE COMPLEXITY
// O(n)

// CODE

class Solution {
public:
    bool solve(TreeNode* root, int tar, string &path){
        if(root == NULL) return false;

        if(root->val == tar) return true;

        path.push_back('L');
        if(solve(root->left, tar, path)) return true;
        path.pop_back();

        path.push_back('R');
        if(solve(root->right, tar, path)) return true;
        path.pop_back();

        return false;
    }
    string getDirections(TreeNode* root, int startValue, int destValue) {
        string RTL = "";
        string RTR = "";
        string res = "";

        solve(root,startValue,RTL);
        solve(root,destValue,RTR);

        int i = 0;
        while(i < RTL.length() && i < RTR.length() && RTL[i] == RTR[i]){
            i++;
        }

        for(int j = 0; j < RTL.length() - i; j++){
            res.push_back('U');
        }

        for(int j = i; j < RTR.length(); j++){
            res.push_back(RTR[j]);
        }
        return res;
    }
};